src/clj/puppetlabs/services/jruby/jruby_puppet_agents.clj

```diff
@@ -65,62 +65,101 @@
     (reset! pool-state new-pool-state)
     (log/info "Swapped JRuby pools, beginning cleanup of old pool.")
-    (doseq [i (range count)]
+    (doseq [i (range old-pool-size)]
       (try
-        (let [id        (inc i)
-              instance  (jruby-internal/borrow-from-pool!*
-                          jruby-internal/borrow-without-timeout-fn
-                          (:pool old-pool))]
+        (let [id (inc i)
+              instance (jruby-internal/borrow-from-pool!*
+                        jruby-internal/borrow-without-timeout-fn
+                        (:pool old-pool))]
           (try
-            (flush-instance! pool-context instance new-pool id config profiler)
+            (jruby-internal/cleanup-pool-instance! instance)
+            (when refill
+              (jruby-internal/create-pool-instance! new-pool id config
+                                                    (partial send-flush-instance! pool-context)
+                                                    profiler)
+              (log/infof "Finished creating JRubyPuppet instance %d of %d"
+                         id old-pool-size))
             (finally
-              (.releaseItem (:pool old-pool) instance false)))
-          (log/infof "Finished creating JRubyPuppet instance %d of %d"
-                     id count))
+              (.releaseItem (:pool old-pool) instance false))))
         (catch Exception e
           (.clear new-pool)
           (.insertPill new-pool (PoisonPill. e))
           (throw (IllegalStateException.
-                   "There was a problem adding a JRubyPuppet instance to the pool."
-                   e)))))
+                  "There was a problem adding a JRubyPuppet instance to the pool."
+                  e)))))
     ;; Add a "RetryPoisonPill" to the pool in case something else is in the
     ;; process of borrowing from the old pool.
     (.insertPill (:pool old-pool) (RetryPoisonPill. (:pool old-pool)))))
 
+(schema/defn ^:always-validate
+  flush-pool-for-shutdown!
+  "Flush of the current JRuby pool when shutting down during a stop."
+  ;; Since this function is only called by the pool-agent, we know that if we
+  ;; receive multiple flush requests before the first one finishes, they will
+  ;; be queued up and we don't need to worry about race conditions between the
+  ;; steps we perform here in the body.
+  [pool-context :- jruby-schemas/PoolContext
+   flush-complete? :- IDeref]
+  (try
+    (log/info "Flush request received; creating new JRuby pool.")
+    (let [{:keys [config pool-state]} pool-context
+          new-pool-state (jruby-internal/create-pool-from-config config)
+          new-pool (:pool new-pool-state)
+          old-pool @pool-state
+          old-pool-size (:size old-pool)]
+      (if-not (pool-initialized? old-pool-size (:pool old-pool))
+        (throw (IllegalStateException. "Attempting to flush a pool that does not appear to have successfully initialized. Aborting.")))
```

Comment:
Do we expect this to happen in production use cases?  What happens if the shutdown is requested via a HUP while the JRubyPool is only partially initialized - say 5 of the 8 expected instances having been initialized?  Will those instances get cleaned up and the service restarted okay?